vapi/vcenter/consumptiondomains/associations/associations.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright (c) 2024-2024 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package associations
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/vmware/govmomi/vapi/rest"
+)
+
+const (
+	basePath         = "/api/vcenter/consumption-domains/zones/cluster"
+	associationsPath = basePath + "/%s/associations"
+)
+
+// Manager extends rest.Client, adding vSphere Zone association related methods.
+type Manager struct {
+	*rest.Client
+}
+
+// NewManager creates a new Manager instance with the given client.
+func NewManager(client *rest.Client) *Manager {
+	return &Manager{
+		Client: client,
+	}
+}
+
+// AddAssociations
+// https://developer.broadcom.com/xapis/vsphere-automation-api/latest/vcenter/api/vcenter/consumption-domains/zones/cluster/zone/associations__action=add/post
+func (c *Manager) AddAssociations(zone string, clusters []string) error {
+	path := c.Resource(fmt.Sprintf(associationsPath, zone)).WithParam("action", "add")
+	req := path.Request(http.MethodPost, clusters)
+	var res map[string]interface{}
```

Comment:
The response [here](https://developer.broadcom.com/xapis/vsphere-automation-api/latest/vcenter/data-structures/ConsumptionDomains_Zones_Cluster_Associations_Status/) has 2 fields. You may want to consider a Status type instead:
```go
// Exception includes error detail for Status responses.
type Exception struct {
	ErrorType string                    `json:"error_type"`
	Messages  []rest.LocalizableMessage `json:"messages,omitempty"`
}

// Status contains information about the outcome of the operation to add a set of
// vSphere clusters to a zone or the operation to remove a set of vSphere clusters
// from a zone.
type Status struct {
	// This is true if the complete operation succeeded without any errors.
	Success bool `json:"success"`
	// FailedClusters represents the associations between the identifiers of the clusters
	// for which the operation failed and the reason of the failure.
	FailedClusters map[string]Exception `json:"failed_clusters,omitempty"`
}
```

Reply:
I know about the second field but I thought I'd keep it simpler and just report the overall status.
I'll follow your advice and add them to the return type since I'll have to revise the diff anyway

Reply:
I re-tested this API

First of all it does not allow for a zone to be associated with more than one cluster. Even the UI uses a radio button that allows single selection.
The bigger problem is that if you provide an incorrect cluster identifier it responds with a message that says the user is not authorized. This is wrong and I will probably log a bug internally.

What I'm going to do is keep the reeturn type and I will change the method signature to accept a single cluster ID instead of an array

Reply:
What about the failed_clusters field, does that not capture some detail when Status == false ?

Multiple clusters might be part of a future enhancement. In which case you could do:
```patch
modified   vapi/vcenter/consumptiondomains/associations/associations.go
@@ -50,9 +50,9 @@ type Status struct {
 
 // AddAssociations
 // https://developer.broadcom.com/xapis/vsphere-automation-api/latest/vcenter/api/vcenter/consumption-domains/zones/cluster/zone/associations__action=add/post
-func (c *Manager) AddAssociations(zone string, cluster string) error {
+func (c *Manager) AddAssociations(zone string, cluster ...string) error {
 	path := c.Resource(fmt.Sprintf(associationsPath, zone)).WithParam("action", "add")
-	req := path.Request(http.MethodPost, []string{cluster})
+	req := path.Request(http.MethodPost, cluster)
 	var res Status
 	if err := c.Do(context.Background(), req, &res); err != nil {
 		return err
```

Reply:
I'd be more than happy to add support for multiple clusters when that is an official feature.
IMO it's best if we provide the same interface as the VC UI which also allows for single-cluster assignment.

As for failed clusters - I mentioned that the API is broken and I don't feel comfortable discussing internal bugs on github. The structure for "failed_clusters" is also undocumented, I can only base my bindings on the buggy response that I get today